### PR TITLE
Refactor regex in providers

### PR DESCRIPTION
**Pull request by eumiro**
Replace `match(r"^…$", …)` with `fullmatch(r"…")`, remove unneeded escaping backslashes, and `search` for text instead of `sub` replacing it.

### CODE REVIEW

File: airflow/providers/amazon/aws/utils/__init__.py [hunk 1]
```diff
@@ -66,8 +66,8 @@ def datetime_to_epoch_us(date_time: datetime) -> int:
 
 
 def get_airflow_version() -> tuple[int, ...]:
-    val = re.sub(r"(\d+\.\d+\.\d+).*", lambda x: x.group(1), version)
-    return tuple(int(x) for x in val.split("."))
+    match = re.match(r"(\d+)\.(\d+)\.(\d+)", version)
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/utils/__init__.py:69]: Not strictly related to this but this should probably be

```suggestion
    match = re.match(r"^(\d+)\.(\d+)\.(\d+)", version)
```

File: airflow/providers/apache/spark/hooks/spark_submit.py [hunk 2]
```diff
@@ -461,31 +461,31 @@ def _process_spark_submit_log(self, itr: Iterator[Any]) -> None:
             # If we run yarn cluster mode, we want to extract the application id from
             # the logs so we can kill the application when we stop it unexpectedly
             if self._is_yarn and self._connection["deploy_mode"] == "cluster":
-                match = re.search("(application[0-9_]+)", line)
+                match = re.search("application[0-9_]+", line)
                 if match:
-                    self._yarn_application_id = match.groups()[0]
+                    self._yarn_application_id = match.group(0)
```
Comment (uranusjr) [inline airflow/providers/apache/spark/hooks/spark_submit.py:466]: Should this be `group(1)` instead?

File: airflow/providers/amazon/aws/utils/__init__.py (hunk 1 above)
Comment (eumiro) [reply airflow/providers/amazon/aws/utils/__init__.py:69]: That's the difference between `re.search` (anywhere in the string), `re.match` (at the beginning) and `re.fullmatch` (both beginning and end). You do not have to add `^` at the beginning of a `re.match` pattern.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Mypy does not like optional groups()

**Review by uranusjr — APPROVED** [review]
I added a commit to fix Mypy

**Review by potiuk — APPROVED** [review]

**Pull request merged**